### ISSUE TITLE
How to handle response from store.request()

### ISSUE THREAD

**Issue by tanj** [infinum/datx#411]
Submitting a:

* [ ] Bug report
* [ ] Feature suggestion
* [X] Support request

--------------

# TLDR

1. Is it possible to use `handleResponse` in my code to handle `store.request` response?
2. Will that response overwrite any manually created relationships that weren't sent to the backend?

# Background
https://github.com/infinum/datx/issues/272#issuecomment-840394424
> 
> 
> Hi @tanj,
> 
> There is no out of the box solution for this (at least for now). The best idea I have is to do a manual API call:
> 
> ```ts
> const rawData = modelToJsonApi(testsheet); // second argument is "onlyDirty" - https://datx.dev/docs/jsonapi/jsonapi-utils#modeltojsonapi
> delete rawData.relationships.report;
> await store.request('/tTestsheet', 'POST', rawData); // https://datx.dev/docs/jsonapi/jsonapi-collection#request
> ```

# How to handle response from store.request()
I was trying to use a manual `request` in another scenario and am running into how do I handle the response data so that my newed up testsheet object gets updated with the response.

I tried handling it the way the `save()` method does by calling `handleResponse`, but I ran into a `Error: Module not found: Error: Can't resolve '@datx/jsonapi/dist/NetworkUtils'` issue when trying to import it.

## The basic flow I'm attempting

- derive a clone by traveling through the relationships newing up objects and copying or setting default values as I go
- as each object is created push it into a `saveQueue`
- when the model is built `concat(saveQueue).subscribe()` after displaying the built model to user.

### Problem 1: auto-generated relationship IDs being submitted
If I try and just do the simplest way of `saveQueue.push(obj.save())` It will try and fail to save because it will include relationships with auto-generated IDs.

### Workaround: Problem 1
To try and work around the auto-generated IDs I was attempting to use a bare `store.request` but I ran into the issue I mentioned above.

### Builder.appendSave()
```ts

    private appendSave<T extends IJsonapiModel>(
        endpoint: string,
        value: T, // This will be all the various relationships passed in from the base object `this.ts`
        exclude: string[]
    ): void {
        // performing raw request because ts.save() would try to save the generated
        // relationships that don't have valid `id`
        // replaces: this.saveQueue.push(this.ts.save());

        const genRecord = new Observable<IRecord>((subscriber) => {
            // @ts-ignore IJsonapiModel.save() signature is different between base and angular
            const rawData = modelToJsonApi(value);
            for (let ex of exclude) {
                if (rawData.relationships && rawData.relationships[ex]) {
                    delete rawData.relationships[ex];
                }
            }
            subscriber.next(rawData);
            subscriber.complete();
        });
        const save = genRecord.pipe(
            tap((obj) => {
                console.log('rawData', obj);
            }),
            concatMap((rawData) =>
                this.tcs.dmc.request(endpoint, 'POST', { data: rawData })
            ),
            map((response) => {
                // Need a way to get `response.data` into `value`
                handleResponse(value) // results in Build Error Module not found
            }),
            tap(() => {
                console.log('save ts', this.ts);
            })
        );
        this.saveQueue.push(save);
    }
```

## Further thoughts
I'm wondering if even if I got the model to sync with the response from the back end that it would just over-write the relationships I built. Say I had a created some `testsheet.report` objects and sent it to the back end with the reports removed. I'd then receive back the response with an empty list for `report.data`

### response data
```json
{
    "data": {
        "attributes": {
            "ixTestsheet": 1747,
            "sInspector": "AM"
        },
        "id": "1747",
        "relationships": {
            "comment": {
                "data": null
            },
            "job": {
                "data": {
                    "id": "120",
                    "type": "tJob"
                }
            },
            "report": {
                "data": []
            },
            "testsheet_type": {
                "data": {
                    "id": "5",
                    "type": "tTestsheetType"
                }
            }
        },
        "type": "tTestsheet"
    },
    "jsonapi": {
        "version": "1.0"
    }
}
```

I think I need to abandon the save queue idea and save each part as it comes or create a new derive endpoint in the backend and take care of it all there. Doing it on the backend is probably much faster anyways...

**Comment by DarkoKukovec** [comment]
Hi @tanj,

Sorry about the delayed response.

I'm guessing you're having the typical json:api issue with creating multiple models at the same time? We're struggling with this also sometimes and while there are multiple ways of doing this (and some might be better suited for specific use cases), the most universal approach we found is to create a custom "form object" model which will contain data from all the models we need to create. The backend will then take this model, create everything it needs on the backend and return the "main" (e.g. tTestsheet) model in the data and other relevant models in the included part. That way there is only a single API call, you get the correct data back from the server right away, and you're not hacking too much around the spec.

The real solution for this would be to implement atomic operations once the json:api v1.1 is published and the libs implement it:
https://github.com/json-api/json-api/blob/3e4b537246ff9310ade08b4bb4ecbbfec391285a/ext/atomic/index.md
